Query to a workflow that was failed by the workflow code produces noisy Error log with not enough context

If a workflow code throws an exception that is failing the workflow, it's getting logged by the worker.
If such a completed with a failure workflow is getting queried, the replay produces the same exact log with an ERROR level and without any additional context that would allow understanding that this exception was effectively "replayed" by the query.

An exception header:

```
2022-08-05 13:22:09.172 [ERROR] 21413 --- [af-5dce1629454d] al.internal.sync.WorkflowExecuteRunnable : Workflow execution failure WorkflowId=entityUrn=urn:indeed:icims-requisition-id:39410&&jobseekerUrn=urn:indeed:icims-candidate-id:testingWithApplicationFailureWithoutAnnotation, RunId=645d1209-6c95-4d06-aeaf-5dce1629454d, WorkflowType=HiringProfessionalWorkflow
io.temporal.failure.ApplicationFailure: message='Unsupported Chart: chartId', type='com.indeed.wayfinder.expedition.host.workflows.hiring.professional.exception.UnsupportedChartException', nonRetryable=true
```

History structure:
```
WorkflowExecutionStarted
WorkflowTaskScheduled
WorkflowTaskStarted
WorkflowTaskCompleted
MarkerRecorded (Version)
MarkerRecorded (Local Activity)
WorkflowExecutionFailed
```

**Describe the solution you'd like**

The original exception should be transformed into Warning from the Error.
The exception that happens in a query-only execution should be on Debug level and contain an additional context that allows to at least track back a query that causes this execution.

--- a/temporal-sdk/src/main/java/io/temporal/failure/TemporalException.java
+++ b/temporal-sdk/src/main/java/io/temporal/failure/TemporalException.java
@@ -26,7 +26,7 @@ package io.temporal.failure;
  * <p>Do not extend by the application code.
  */
 public class TemporalException extends RuntimeException {
-  public TemporalException(String message, Throwable cause) {
+  protected TemporalException(String message, Throwable cause) {
     super(message, cause, false, true);
   }
 }

--- a/temporal-sdk/src/main/java/io/temporal/failure/TemporalFailure.java
+++ b/temporal-sdk/src/main/java/io/temporal/failure/TemporalFailure.java
@@ -42,7 +42,7 @@ public abstract class TemporalFailure extends TemporalException {
   private Optional<Failure> failure = Optional.empty();
   private final String originalMessage;
 
-  protected TemporalFailure(String message, String originalMessage, Throwable cause) {
+  TemporalFailure(String message, String originalMessage, Throwable cause) {
     super(message, cause);
     this.originalMessage = originalMessage;
   }

--- a/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowContext.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowContext.java
@@ -56,7 +56,9 @@ public interface ReplayWorkflowContext extends ReplayAware {
 
   WorkflowType getWorkflowType();
 
-  /** Is cancel of the workflow requested. */
+  /**
+   * @return true if cancellation of the workflow is requested.
+   */
   boolean isCancelRequested();
 
   /**
@@ -279,4 +281,12 @@ public interface ReplayWorkflowContext extends ReplayAware {
    * @return workflow cron schedule
    */
   String getCronSchedule();
+
+  /**
+   * This method is mostly used to decrease logging verbosity for replay-only scenarios.
+   *
+   * @return query name if an execution is a full replay caused by a direct query, null otherwise
+   */
+  @Nullable
+  String getFullReplayDirectQueryName();
 }

--- a/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowContextImpl.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowContextImpl.java
@@ -52,24 +52,30 @@ import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 /**
- * TODO(maxim): callbacks usage is non consistent. It accepts Optional and Exception which can be
- * null. Either switch both to Optional or both to nullable.
+ * TODO callbacks usage is non consistent. It accepts Optional and Exception which can be null.
+ * Switch both to nullable.
  */
 final class ReplayWorkflowContextImpl implements ReplayWorkflowContext {
 
-  private final WorkflowContext workflowContext;
-  private final Scope metricsScope;
-  private final boolean enableLoggingInReplay;
   private final WorkflowStateMachines workflowStateMachines;
+  private final WorkflowContext workflowContext;
+  private final @Nullable String fullReplayDirectQueryName;
+  private final Scope replayAwareWorkflowMetricsScope;
+  private final SingleWorkerOptions workerOptions;
 
+  /**
+   * @param fullReplayDirectQueryName query name if an execution is a full replay caused by a direct
+   *     query, null otherwise
+   */
   ReplayWorkflowContextImpl(
       WorkflowStateMachines workflowStateMachines,
       String namespace,
       WorkflowExecutionStartedEventAttributes startedAttributes,
       WorkflowExecution workflowExecution,
       long runStartedTimestampMillis,
-      SingleWorkerOptions options,
-      Scope metricsScope) {
+      @Nullable String fullReplayDirectQueryName,
+      SingleWorkerOptions workerOptions,
+      Scope workflowMetricsScope) {
     this.workflowStateMachines = workflowStateMachines;
     this.workflowContext =
         new WorkflowContext(
@@ -77,15 +83,16 @@ final class ReplayWorkflowContextImpl implements ReplayWorkflowContext {
             workflowExecution,
             startedAttributes,
             runStartedTimestampMillis,
-            options.getContextPropagators());
-    this.enableLoggingInReplay = options.getEnableLoggingInReplay();
-    this.metricsScope =
-        new ReplayAwareScope(metricsScope, this, workflowStateMachines::currentTimeMillis);
+            workerOptions.getContextPropagators());
+    this.fullReplayDirectQueryName = fullReplayDirectQueryName;
+    this.replayAwareWorkflowMetricsScope =
+        new ReplayAwareScope(workflowMetricsScope, this, workflowStateMachines::currentTimeMillis);
+    this.workerOptions = workerOptions;
   }
 
   @Override
   public boolean getEnableLoggingInReplay() {
-    return enableLoggingInReplay;
+    return workerOptions.getEnableLoggingInReplay();
   }
 
   @Override
@@ -100,7 +107,7 @@ final class ReplayWorkflowContextImpl implements ReplayWorkflowContext {
 
   @Override
   public Scope getMetricsScope() {
-    return metricsScope;
+    return replayAwareWorkflowMetricsScope;
   }
 
   @Nonnull
@@ -339,5 +346,11 @@ final class ReplayWorkflowContextImpl implements ReplayWorkflowContext {
   @Override
   public String getCronSchedule() {
     return workflowContext.getCronSchedule();
+  }
+
+  @Nullable
+  @Override
+  public String getFullReplayDirectQueryName() {
+    return fullReplayDirectQueryName;
   }
 }

--- a/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java
@@ -95,7 +95,7 @@ class ReplayWorkflowRunTaskHandler implements WorkflowRunTaskHandler {
       String namespace,
       ReplayWorkflow workflow,
       PollWorkflowTaskQueueResponseOrBuilder workflowTask,
-      SingleWorkerOptions options,
+      SingleWorkerOptions workerOptions,
       Scope metricsScope,
       BiFunction<LocalActivityTask, Duration, Boolean> localActivityTaskPoller) {
     this.service = service;
@@ -113,6 +113,8 @@ class ReplayWorkflowRunTaskHandler implements WorkflowRunTaskHandler {
     this.workflow = workflow;
 
     this.workflowStateMachines = new WorkflowStateMachines(new StatesMachinesCallbackImpl());
+    String fullReplayDirectQueryType =
+        workflowTask.hasQuery() ? workflowTask.getQuery().getQueryType() : null;
     ReplayWorkflowContextImpl context =
         new ReplayWorkflowContextImpl(
             workflowStateMachines,
@@ -120,7 +122,8 @@ class ReplayWorkflowRunTaskHandler implements WorkflowRunTaskHandler {
             startedEvent,
             workflowTask.getWorkflowExecution(),
             Timestamps.toMillis(firstEvent.getEventTime()),
-            options,
+            fullReplayDirectQueryType,
+            workerOptions,
             metricsScope);
 
     this.replayWorkflowExecutor =

--- a/temporal-sdk/src/main/java/io/temporal/internal/replay/WorkflowContext.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/replay/WorkflowContext.java
@@ -40,15 +40,16 @@ import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 final class WorkflowContext {
-
   private final long runStartedTimestampMillis;
-  private boolean cancelRequested;
-  private ContinueAsNewWorkflowExecutionCommandAttributes continueAsNewOnCompletion;
   private final WorkflowExecutionStartedEventAttributes startedAttributes;
   private final String namespace;
-  private SearchAttributes.Builder searchAttributes;
   private final List<ContextPropagator> contextPropagators;
   @Nonnull private final WorkflowExecution workflowExecution;
+
+  // Mutable, accumulated during execution:
+  private SearchAttributes.Builder searchAttributes;
+  private boolean cancelRequested;
+  private ContinueAsNewWorkflowExecutionCommandAttributes continueAsNewOnCompletion;
 
   WorkflowContext(
       String namespace,

--- a/temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowExecuteRunnable.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowExecuteRunnable.java
@@ -28,12 +28,14 @@ import io.temporal.api.history.v1.WorkflowExecutionStartedEventAttributes;
 import io.temporal.common.interceptors.Header;
 import io.temporal.failure.FailureConverter;
 import io.temporal.failure.TemporalFailure;
+import io.temporal.internal.replay.ReplayWorkflowContext;
 import io.temporal.internal.worker.WorkflowExecutionException;
 import io.temporal.worker.WorkflowImplementationOptions;
 import io.temporal.workflow.Workflow;
 import io.temporal.workflow.WorkflowInfo;
 import java.util.Objects;
 import java.util.Optional;
+import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -75,22 +77,11 @@ class WorkflowExecuteRunnable implements Runnable {
       Class<? extends Throwable>[] failTypes =
           implementationOptions.getFailWorkflowExceptionTypes();
       if (exception instanceof TemporalFailure) {
-        logWorkflowExecutionException(Workflow.getInfo(), exception);
-        throw new WorkflowExecutionException(
-            FailureConverter.exceptionToFailure(exception, context.getDataConverter()));
+        throwAndFailWorkflowExecution(exception);
       }
       for (Class<? extends Throwable> failType : failTypes) {
         if (failType.isAssignableFrom(exception.getClass())) {
-          // fail workflow
-          if (log.isErrorEnabled()) {
-            boolean cancelRequested =
-                WorkflowInternal.getRootWorkflowContext().getContext().isCancelRequested();
-            if (!cancelRequested || !FailureConverter.isCanceledCause(exception)) {
-              logWorkflowExecutionException(Workflow.getInfo(), exception);
-            }
-          }
-          throw new WorkflowExecutionException(
-              FailureConverter.exceptionToFailure(exception, context.getDataConverter()));
+          throwAndFailWorkflowExecution(exception);
         }
       }
       throw wrap(exception);
@@ -119,15 +110,44 @@ class WorkflowExecuteRunnable implements Runnable {
     return context.handleQuery(type, args);
   }
 
-  private void logWorkflowExecutionException(WorkflowInfo info, Throwable exception) {
-    log.error(
-        "Workflow execution failure "
-            + "WorkflowId="
-            + info.getWorkflowId()
-            + ", RunId="
-            + info.getRunId()
-            + ", WorkflowType="
-            + info.getWorkflowType(),
-        exception);
+  private void throwAndFailWorkflowExecution(Throwable exception) {
+    ReplayWorkflowContext replayWorkflowContext = context.getContext();
+    @Nullable
+    String fullReplayDirectQueryName = replayWorkflowContext.getFullReplayDirectQueryName();
+    WorkflowInfo info = Workflow.getInfo();
+
+    if (fullReplayDirectQueryName != null) {
+      if (log.isDebugEnabled()
+          && !requestedCancellation(replayWorkflowContext.isCancelRequested(), exception)) {
+        log.debug(
+            "Replayed workflow execution failure WorkflowId='{}', RunId={}, WorkflowType='{}' for direct query QueryType='{}'",
+            info.getWorkflowId(),
+            info.getRunId(),
+            info.getWorkflowType(),
+            fullReplayDirectQueryName,
+            exception);
+      }
+    } else {
+      if (log.isWarnEnabled()
+          && !requestedCancellation(replayWorkflowContext.isCancelRequested(), exception)) {
+        log.warn(
+            "Workflow execution failure WorkflowId='{}', RunId={}, WorkflowType='{}'",
+            info.getWorkflowId(),
+            info.getRunId(),
+            info.getWorkflowType(),
+            exception);
+      }
+    }
+
+    throw new WorkflowExecutionException(
+        FailureConverter.exceptionToFailure(exception, context.getDataConverter()));
+  }
+
+  /**
+   * @return true if both workflow cancellation is requested and the exception contains a
+   *     cancellation exception in the chain
+   */
+  private boolean requestedCancellation(boolean cancelRequested, Throwable exception) {
+    return cancelRequested && FailureConverter.isCanceledCause(exception);
   }
 }

--- a/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/DirectQueryReplaysDontSpamLogWithWorkflowExecutionExceptionsTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/DirectQueryReplaysDontSpamLogWithWorkflowExecutionExceptionsTest.java
@@ -1,0 +1,95 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.workflow.queryTests;
+
+import static org.junit.Assert.*;
+
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
+import io.temporal.client.WorkflowException;
+import io.temporal.failure.ApplicationFailure;
+import io.temporal.internal.Issue;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.workflow.shared.TestWorkflows;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Verifies that full replays of failed workflows for queries don't spam log with misleading
+ * workflow exceptions that look like original workflow execution exceptions.
+ */
+@Issue("https://github.com/temporalio/sdk-java/issues/1348")
+public class DirectQueryReplaysDontSpamLogWithWorkflowExecutionExceptionsTest {
+
+  private static final AtomicInteger workflowCodeExecutionCount = new AtomicInteger();
+  private final ListAppender<ILoggingEvent> workflowExecuteRunnableLoggerAppender =
+      new ListAppender<>();
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder().setWorkflowTypes(TestWorkflowNonRetryableFlag.class).build();
+
+  @Before
+  public void setUp() throws Exception {
+    workflowCodeExecutionCount.set(0);
+
+    Logger workflowExecuteRunnableLogger =
+        (Logger) LoggerFactory.getLogger("io.temporal.internal.sync.WorkflowExecuteRunnable");
+    workflowExecuteRunnableLoggerAppender.start();
+    workflowExecuteRunnableLogger.addAppender(workflowExecuteRunnableLoggerAppender);
+  }
+
+  @Test
+  public void queriedWorkflowFailureDoesntProduceAdditionalLogs() {
+    TestWorkflows.TestWorkflowWithQuery testWorkflow =
+        testWorkflowRule.newWorkflowStub(TestWorkflows.TestWorkflowWithQuery.class);
+    assertThrows(WorkflowException.class, testWorkflow::execute);
+    assertEquals(
+        "Workflow execution exception should be logged",
+        1,
+        workflowExecuteRunnableLoggerAppender.list.size());
+    testWorkflow.query();
+    assertEquals(2, workflowCodeExecutionCount.get());
+    assertEquals(
+        "There was two execution - one original and one full replay for query. "
+            + "But only one original exception should be logged.",
+        1,
+        workflowExecuteRunnableLoggerAppender.list.size());
+  }
+
+  public static class TestWorkflowNonRetryableFlag implements TestWorkflows.TestWorkflowWithQuery {
+
+    @Override
+    public String execute() {
+      workflowCodeExecutionCount.incrementAndGet();
+      throw ApplicationFailure.newNonRetryableFailure("SomeMessage", "SomeType");
+    }
+
+    @Override
+    public String query() {
+      return null;
+    }
+  }
+}

--- a/temporal-testing/src/main/java/io/temporal/internal/sync/DummySyncWorkflowContext.java
+++ b/temporal-testing/src/main/java/io/temporal/internal/sync/DummySyncWorkflowContext.java
@@ -40,6 +40,7 @@ import io.temporal.internal.replay.StartChildWorkflowExecutionParameters;
 import io.temporal.workflow.Functions;
 import java.time.Duration;
 import java.util.*;
+import javax.annotation.Nullable;
 
 public class DummySyncWorkflowContext {
   public static SyncWorkflowContext newDummySyncWorkflowContext() {
@@ -279,6 +280,12 @@ public class DummySyncWorkflowContext {
     @Override
     public String getCronSchedule() {
       return "dummy-cron-schedule";
+    }
+
+    @Nullable
+    @Override
+    public String getFullReplayDirectQueryName() {
+      return null;
     }
   }
 }

